From f480b04528fc7b66e8d9b4d3d1eb1ab43ed41e00 Mon Sep 17 00:00:00 2001
From: Nan Lin <linnan@chromium.org>
Date: Mon, 21 Oct 2024 18:35:56 +0000
Subject: [PATCH] Replace BrowsingDataFilterBuilderImpl::MatchesWithSavedStorageKey with std::equal_to

This is to avoid dangling pointer when the callback is run.

Bug: 353506686
Change-Id: I2dae05aa5b4c255ee7ffed5b589cd6eb64df34dc
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5947364
Auto-Submit: Nan Lin <linnan@chromium.org>
Commit-Queue: Avi Drissman <avi@chromium.org>
Reviewed-by: Christian Dullweber <dullweber@chromium.org>
Reviewed-by: Avi Drissman <avi@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1371535}
---

--- a/content/browser/browsing_data/browsing_data_filter_builder_impl.cc
+++ b/content/browser/browsing_data/browsing_data_filter_builder_impl.cc
@@ -4,6 +4,7 @@
 
 #include "content/browser/browsing_data/browsing_data_filter_builder_impl.h"
 
+#include <functional>
 #include <string>
 #include <vector>
 
@@ -212,12 +213,6 @@
   return storage_key_.has_value();
 }
 
-bool BrowsingDataFilterBuilderImpl::MatchesWithSavedStorageKey(
-    const blink::StorageKey& other_key) const {
-  DCHECK(storage_key_.has_value());
-  return storage_key_.value() == other_key;
-}
-
 bool BrowsingDataFilterBuilderImpl::MatchesAllOriginsAndDomains() {
   return MatchesMostOriginsAndDomains() && origins_.empty() &&
          domains_.empty() && cookie_partition_key_collection_.ContainsAllKeys();
@@ -266,9 +261,8 @@
   if (HasStorageKey()) {
     CHECK(StorageKeyInCookiePartitionKeyCollection(
         *storage_key_, cookie_partition_key_collection_));
-    return base::BindRepeating(
-        &BrowsingDataFilterBuilderImpl::MatchesWithSavedStorageKey,
-        base::Unretained(this));
+    return base::BindRepeating(std::equal_to<const blink::StorageKey&>(),
+                               *storage_key_);
   }
   return base::BindRepeating(&MatchesStorageKey, origins_, domains_, mode_,
                              origin_mode_);
--- a/content/browser/browsing_data/browsing_data_filter_builder_impl.h
+++ b/content/browser/browsing_data/browsing_data_filter_builder_impl.h
@@ -39,8 +39,6 @@
   void SetStorageKey(
       const std::optional<blink::StorageKey>& storage_key) override;
   bool HasStorageKey() const override;
-  bool MatchesWithSavedStorageKey(
-      const blink::StorageKey& other_key) const override;
   bool MatchesAllOriginsAndDomains() override;
   bool MatchesMostOriginsAndDomains() override;
   bool MatchesNothing() override;
--- a/content/public/browser/browsing_data_filter_builder.h
+++ b/content/public/browser/browsing_data_filter_builder.h
@@ -106,11 +106,6 @@
   // Returns whether the StorageKey is set (e.g. using the method above).
   virtual bool HasStorageKey() const = 0;
 
-  // Returns whether the filter's StorageKey matches the given one.
-  // Note: the StorageKey in the filter has to be set.
-  virtual bool MatchesWithSavedStorageKey(
-      const blink::StorageKey& other_key) const = 0;
-
   // Returns true if we're an empty preserve list, where we delete everything.
   virtual bool MatchesAllOriginsAndDomains() = 0;
 
